### ISSUE TITLE
Hints can be lost during topology change or tablet transition

### ISSUE THREAD

**Issue by piodul** [scylladb/scylladb#19835]
When hinted handoff replays a hint, it considers two possibilities:

- If the destination node is a replica ("natural endpoint"), send the hint directly to it.
- If the destination node is not a replica (not a "natural endpoint"), send the hint to all current replicas.

However, during topology operations in vnode-backed tables and during tablet migrations this can lead to the loss of the hint. This is because the replica that the hint was directly replayed to might no longer be a replica in a while, and the process of streaming data to the new replica might miss it. In case of normal writes we do "double writes", i.e. send writes to the union of the sets of old and new replicas; perhaps we should consider writing to the pending replicas if the replica will go away after streaming.

Here is the relevant code from the hinted handoff logic:

```c++
        inet_address_vector_replica_set natural_endpoints = erm->get_natural_endpoints(std::move(token));

        // ...

        if (maybe_addr && boost::range::find(natural_endpoints, *maybe_addr) != natural_endpoints.end()) {
            manager_logger.trace("Sending directly to {}", end_point_key());
            return _proxy.send_hint_to_endpoint(std::move(m), std::move(ermp), *maybe_addr);
        } else {
            manager_logger.trace("Endpoints set has changed and {} is no longer a replica. Mutating from scratch...", end_point_key());
            return _proxy.send_hint_to_all_replicas(std::move(m));
        }
```

**Comment by michoecho** [comment]
I don't understand this issue. I suspect that there actually is no issue.

Could you give some example step-by-step scenario of the problem you are thinking about?

**Comment by bhalevy** [comment]
> * If the destination node is not a replica (not a "natural endpoint"), send the hint to all current replicas.

@michoecho here's an example of my understanding:
- Current replica set is [A, B, C]
- write succeeds on [A, B], and a hint is logged for node C
- before the hint is replayed, D bootstraps and the token moves from C to D
- hint is replayed to node C while D is pending, but it's too late, since streaming for that token is already done
- C is cleaned up, replayed data is lost, and D has a stale copy until next repair.

**Comment by bhalevy** [comment]
Another example, that may explain https://github.com/scylladb/scylla-dtest/issues/4582:
- Current replica set is [A, B, C]
- write succeeds on [A, B], and a hint is logged for node C
- before the hint is replayed, C is decommissioned and the token migrates from C to D
- hint is replayed to node C while C is pending leave (being decommissioned), but it's too late, since streaming for that token is already done
- C complete decommissioning, replayed data is lost, and D has a stale copy until next repair.

**Comment by bhalevy** [comment]
Since hints are lossy by design I'm not sure that the `bug` label applies here.

**Comment by mykaul** [comment]
This does look like a serious (?) gap in tablets functionality (vs. vnodes?), what's the next step here?

**Comment by michoecho** [comment]
> This does look like a serious (?) gap in tablets functionality

But this isn't specific to tablets. It applies to every topology change.

**Comment by mykaul** [comment]
> > This does look like a serious (?) gap in tablets functionality
> 
> But this isn't specific to tablets. It applies to every topology change.

But I assume (perhaps wrongfully) they are relatively rare  in vnodes compared to tablet load balancing?

**Comment by michoecho** [comment]
> > > This does look like a serious (?) gap in tablets functionality
> > 
> > 
> > But this isn't specific to tablets. It applies to every topology change.
> 
> But I assume (perhaps wrongfully) they are relatively rare in vnodes compared to tablet load balancing?

Not really, why? Topology operations happen whenever you add or remove a node. This doesn't depend on whether you are using tablets or vnodes.

**Comment by mykaul** [comment]
> > > > This does look like a serious (?) gap in tablets functionality
> > > 
> > > 
> > > But this isn't specific to tablets. It applies to every topology change.
> > 
> > 
> > But I assume (perhaps wrongfully) they are relatively rare in vnodes compared to tablet load balancing?
> 
> Not really, why? Topology operations happen whenever you add or remove a node. This doesn't depend on whether you are using tablets or vnodes.

OK - so it won't happen on load balancing (across shards) ? Since hints are per node, not per shard?

**Comment by swasik** [comment]
@piodul with https://github.com/scylladb/scylladb/pull/20488 merged and backported is there still something to be done to solve this issue?

**Comment by mykaul** [comment]
> [@piodul](https://github.com/piodul) with [#20488](https://github.com/scylladb/scylladb/pull/20488) merged and backported is there still something to be done to solve this issue?

@piodul ?

**Comment by tgrabiec** [comment]
I think tablet migration is not handled correctly:

```
future<> hint_sender::send_one_mutation(frozen_mutation_and_schema m) {
...
        if (std::ranges::contains(natural_endpoints, dst) && !tm.is_leaving(dst)) {
            manager_logger.trace("Sending directly to {}", dst);
            return _proxy.send_hint_to_endpoint(std::move(m), std::move(ermp), dst);
        } else {
            if (manager_logger.is_enabled(log_level::trace)) {
                if (tm.is_leaving(end_point_key())) {
                    manager_logger.trace("The original target endpoint {} is leaving. Mutating from scratch...", dst);
                } else {
                    manager_logger.trace("Endpoints set has changed and {} is no longer a replica. Mutating from scratch...", dst);
                }
            }
            return _proxy.send_hint_to_all_replicas(std::move(m));
        }

```

The is_leaving() check only takes care of decommission, but doesn't recognize the case when one of the tablet replicas is leaving, which should result in send_hint_to_all_replicas() for the same reasons as when the node is leaving. If this hint targets the leaving replica and streaming already went past the key, we will lose that hint after migration.

**Comment by raphaelsc** [comment]
> I think tablet migration is not handled correctly:
> 
> ```
> future<> hint_sender::send_one_mutation(frozen_mutation_and_schema m) {
> ...
>         if (std::ranges::contains(natural_endpoints, dst) && !tm.is_leaving(dst)) {
>             manager_logger.trace("Sending directly to {}", dst);
>             return _proxy.send_hint_to_endpoint(std::move(m), std::move(ermp), dst);
>         } else {
>             if (manager_logger.is_enabled(log_level::trace)) {
>                 if (tm.is_leaving(end_point_key())) {
>                     manager_logger.trace("The original target endpoint {} is leaving. Mutating from scratch...", dst);
>                 } else {
>                     manager_logger.trace("Endpoints set has changed and {} is no longer a replica. Mutating from scratch...", dst);
>                 }
>             }
>             return _proxy.send_hint_to_all_replicas(std::move(m));
>         }
> ```
> 
> The is_leaving() check only takes care of decommission, but doesn't recognize the case when one of the tablet replicas is leaving, which should result in send_hint_to_all_replicas() for the same reasons as when the node is leaving. If this hint targets the leaving replica and streaming already went past the key, we will lose that hint after migration.

This is a potential source of data resurrection, since the lost hint might be a tombstone.

I think a better fix is to include only the pending, to avoid sending redundant data as much as possible. Of course, it might happen that hint destination is not the leaving replica, and we sent hint unecessarily to pending, but it's better than sending hints to all replicas during migrations. Later, we can possibly add a get_leaving_replicas() and then we only stream to pending if hint dest is leaving.

```
diff --git a/service/storage_proxy.hh b/service/storage_proxy.hh
index b969bf67c1..fb1d276ed5 100644
--- a/service/storage_proxy.hh
+++ b/service/storage_proxy.hh
@@ -641,7 +641,7 @@ class storage_proxy : public seastar::async_sharded_service<storage_proxy>, publ
     // Send a mutation to a specific remote target as a hint.
     // Unlike regular mutations during write operations, hints are sent on the streaming connection
     // and use different RPC verb.
-    future<> send_hint_to_endpoint(frozen_mutation_and_schema fm_a_s, locator::effective_replication_map_ptr ermp, locator::host_id target);
+    future<> send_hint_to_endpoint(frozen_mutation_and_schema fm_a_s, locator::effective_replication_map_ptr ermp, locator::host_id target, host_id_vector_topology_change pending_endpoints);
 
     /**
      * Performs the truncate operatoin, which effectively deletes all data from
diff --git a/db/hints/internal/hint_sender.cc b/db/hints/internal/hint_sender.cc
index ed8dbb901e..40f1965a72 100644
--- a/db/hints/internal/hint_sender.cc
+++ b/db/hints/internal/hint_sender.cc
@@ -248,9 +248,10 @@ void hint_sender::start() {
 future<> hint_sender::send_one_mutation(frozen_mutation_and_schema m) {
     auto ermp = _db.find_column_family(m.s).get_effective_replication_map();
     auto token = dht::get_token(*m.s, m.fm.key());
-    host_id_vector_replica_set natural_endpoints = ermp->get_natural_replicas(std::move(token));
+    host_id_vector_replica_set natural_endpoints = ermp->get_natural_replicas(token);
+    host_id_vector_topology_change pending_endpoints = erm->get_pending_replicas(token);
 
-    return futurize_invoke([this, m = std::move(m), ermp = std::move(ermp), &natural_endpoints] () mutable -> future<> {
+    return futurize_invoke([this, m = std::move(m), ermp = std::move(ermp), &natural_endpoints, &pending_endpoints] () mutable -> future<> {
         // The fact that we send with CL::ALL in both cases below ensures that new hints are not going
         // to be generated as a result of hints sending.
         const auto& tm = ermp->get_token_metadata();
@@ -258,7 +259,7 @@ future<> hint_sender::send_one_mutation(frozen_mutation_and_schema m) {
 
         if (std::ranges::contains(natural_endpoints, dst) && !tm.is_leaving(dst)) {
             manager_logger.trace("Sending directly to {}", dst);
-            return _proxy.send_hint_to_endpoint(std::move(m), std::move(ermp), dst);
+            return _proxy.send_hint_to_endpoint(std::move(m), std::move(ermp), dst, std::move(pending_endpoints));
         } else {
             if (manager_logger.is_enabled(log_level::trace)) {
                 if (tm.is_leaving(end_point_key())) {
diff --git a/service/storage_proxy.cc b/service/storage_proxy.cc
index 762ff04449..3cda06cf9c 100644
--- a/service/storage_proxy.cc
+++ b/service/storage_proxy.cc
@@ -4225,12 +4225,12 @@ future<> storage_proxy::send_to_endpoint(
             cancellable);
 }
 
-future<> storage_proxy::send_hint_to_endpoint(frozen_mutation_and_schema fm_a_s, locator::effective_replication_map_ptr ermp, locator::host_id target) {
+future<> storage_proxy::send_hint_to_endpoint(frozen_mutation_and_schema fm_a_s, locator::effective_replication_map_ptr ermp, locator::host_id target, host_id_vector_topology_change pending_endpoints) {
     return send_to_endpoint(
             std::make_unique<hint_mutation>(std::move(fm_a_s)),
             std::move(ermp),
             std::move(target),
-            host_id_vector_topology_change{},
+            std::move(pending_endpoints),
             db::write_type::SIMPLE,
             tracing::trace_state_ptr(),
             get_stats(),
```

**Comment by wmitros** [comment]
I agree with @raphaelsc's idea, so I sent a patch: https://github.com/scylladb/scylladb/pull/25590. The suggested patch looks good so I basically copied it. And @bhalevy's example is very useful for clarifying the issue, so I included it as well
